#!/usr/bin/env python

from itertools import cycle

import numpy as np
from scipy import stats
import pandas as pd
import matplotlib
from matplotlib import pyplot as plt

import click

from pandarallel import pandarallel

pandarallel.initialize(verbose=0)


@click.group()
def cli():
    pass


def load_files(genotypefile, phenotypefile):
    with open(genotypefile, "r") as gfile, open(phenotypefile, "r") as pfile:

        genotypes_df = pd.read_csv(gfile, index_col=("Chromosome", "Coordinate"))
        phenotypes_df = pd.read_csv(pfile, index_col="Sample_Name")

        # reindexing makes phenotype sample order conform to genotype sample order
        # sample names in genotype header that don't exist get a NaN in the
        # phenotype columns; sample names in phenotypes that don't exist in
        # genotype header are dropped
        phenotypes_df = phenotypes_df.reindex(genotypes_df.columns)

        return genotypes_df, phenotypes_df


assoc_test_dict = {
    "anova": stats.f_oneway,
    "alexander": stats.alexandergovern,
    "kruskal": stats.kruskal,
}


def assoc_by_genotype(genorow, phenodf, focalpheno, genos=(0, 1), test="anova"):
    groups = [
        phenodf[genorow == i][focalpheno].dropna().values.flatten() for i in genos
    ]
    statistic, pval = assoc_test_dict[test](*groups)
    return statistic, pval

def phenomean_by_genotype(genorow, phenodf, focalpheno, genos=(0, 1)):
    groups = [
        phenodf[genorow == i][focalpheno].dropna().values.flatten() for i in genos
    ]
    group_means = [g.mean() for g in groups]
    return group_means


def assoc_test_pandarallel(genodf, phenodf, focalpheno, genos=(0, 1), test="anova"):
    assoc = genodf.parallel_apply(
        assoc_by_genotype,
        args=(phenodf, focalpheno, genos, test),
        axis=1,
        result_type="expand",
    )
    assoc.columns = ("statistic", "Pvalue")
    assoc["log10Pvalue"] = -np.log10(assoc.Pvalue)
    return assoc


def assoc_test_pandas(genodf, phenodf, focalpheno, genos=(0, 1), test="anova"):
    assoc = genodf.apply(
        assoc_by_genotype,
        args=(phenodf, focalpheno, genos, test),
        axis=1,
        result_type="expand",
    )
    assoc.columns = ("statistic", "Pvalue")
    assoc["log10Pvalue"] = -np.log10(assoc.Pvalue)
    return assoc


def means_by_genotype(genodf, phenodf, focalpheno, genos=(0, 1), test="anova"):
    means = genodf.parallel_apply(
        phenomean_by_genotype,
        args=(phenodf, focalpheno, genos),
        axis=1,
        result_type="expand",
    )
    geno_name_dict = {
        (0,1): ["ref_mean", "alt_mean"],
        (-1,1): ["ref_mean", "alt_mean"],
        (0,1,2): ["ref_mean", "het_mean" "alt_mean"],
    }

    means.columns = geno_name_dict[genos]
    return means


@cli.command()
@click.option(
    "--threshold",
    type=float,
    default=None,
    help=("-log10 pvalue at which to draw the significance threshold.")
)
@click.option(
    "--rotlabels",
    is_flag = True,
)
@click.option(
    "--ylim",
    type=float,
    default=None,
    help=(
        "Specifies maximum -log10 pvalue of the y-axis for plotting consistency. "
    )
)
@click.option(
    "--output",
    "-o",
    type=str,
    default="",
    help=(
        "Specifies an output file to which a PNG image of the generated "
        "figure will be written. If unspecified, an interactive matplotlib figure "
        "will be shown instead and a save file can be generated from within "
        "this window."
    ),
)
@click.argument("statfile", type=click.File("r"))
@click.argument("chromfile", type=click.File("r"))
def plot(statfile, chromfile, output, ylim, threshold, rotlabels):
    """
    Create a "manhattan" style plot showing the -log(p-values) per site.

    STATFILE is a file generated by the 'quickQTL.py stats' command.

    CHROMFILE is a CSV file giving length of each chromosome.
    See README for formatting requirements.
    """
    association_stats = pd.read_csv(statfile)
    chroms = pd.read_csv(chromfile, na_values = ".")

    chroms = chroms.set_index("Chromosome")
    chroms["Offset"] = [0] + chroms.Length[:-1].cumsum().values.tolist()
    chroms["Midpoint"] = (chroms.Length + 2 * chroms.Offset) / 2
    chroms["Endpoint"] = chroms.Length + chroms.Offset

    grouped_stats = association_stats.groupby("Chromosome")

    if bool(output):
        matplotlib.use("agg")

    colorcycle = cycle(plt.cm.Dark2.colors)
    fig, ax = plt.subplots(1, 1)

    maxX = 0
    for chrom in chroms.index:
        grp = grouped_stats.get_group(chrom)
        offset = chroms.loc[chrom, "Offset"]
        ax.plot(
            grp.Coordinate + offset,
            grp.log10Pvalue,
            markersize=2,
            marker="o",
            linestyle="None",
            alpha=0.75,
            color=next(colorcycle),
        )
        if (grp.Coordinate.max() + offset) > maxX:
            maxX = grp.Coordinate.max() + offset
    ax.set_xticks(chroms.Midpoint, labels=chroms.index)
    ax.set_xticks(chroms.Endpoint, minor=True)
    ax.set_xlabel("Coordinates")
    ax.set_ylabel(r"-$\log_{10}$(p-value)")
    if ylim:
        ax.set_ylim(0, ylim)

    if threshold:
        ax.hlines(threshold, 0, maxX, 
                  alpha = 0.5,
                  color="black", linestyles="dashed")

    if rotlabels:
        ax.tick_params(axis='x', labelrotation=45)

    if bool(output):
        plt.savefig(output, format="png", dpi=150)
    else:
        plt.show()


@cli.command()
@click.option(
    "--output",
    "-o",
    type=str,
    default="",
    help=(
        "Specifies an output file to which a PNG image of the generated "
        "figure will be written. If unspecified, an interactive matplotlib figure "
        "will be shown instead and a save file can be generated from within "
        "this window."
    ),
)
@click.option(
    "--diff",
    is_flag = True,
    help=(
        "Plot mean difference, alt_mean - ref_mean, instead of absolute means."
    )
)
@click.argument("phenomeanfile", type=click.File("r"))
@click.argument("chromfile", type=click.File("r"))
def plot_phenotype(phenomeanfile, chromfile, output, diff):
    """
    Create a "manhattan" style plot showing the allele-specific means per site.

    PHENOMEANFILE is a file generated by the 'quickQTL.py phenotype' command.

    CHROMFILE is a CSV file giving length of each chromosome.
    See README for formatting requirements.
    """
    association_stats = pd.read_csv(phenomeanfile)
    chroms = pd.read_csv(chromfile, na_values = ".")

    chroms = chroms.set_index("Chromosome")
    chroms["Offset"] = [0] + chroms.Length[:-1].cumsum().values.tolist()
    chroms["Midpoint"] = (chroms.Length + 2 * chroms.Offset) / 2
    chroms["Endpoint"] = chroms.Length + chroms.Offset

    grouped_stats = association_stats.groupby("Chromosome")

    if bool(output):
        matplotlib.use("agg")

    colorcycle = cycle(plt.cm.Dark2.colors)
    fig, ax = plt.subplots(1, 1)

    ct = 0
    for chrom in chroms.index:
        grp = grouped_stats.get_group(chrom)
        clr = color=next(colorcycle)
        if diff:
            ax.plot(
                grp.Coordinate + chroms.loc[chrom, "Offset"],
                grp.alt_mean - grp.ref_mean,
                markersize=2,
                marker="o",
                linestyle="None",
                alpha=0.75,
                color=clr,
            )            

        else:
            ax.plot(
                grp.Coordinate + chroms.loc[chrom, "Offset"],
                grp.ref_mean,
                markersize=4,
                marker="o",
                fillstyle="none",
                linestyle="None",
                alpha=0.75,
                color=clr,
            )
            ax.plot(
                grp.Coordinate + chroms.loc[chrom, "Offset"],
                grp.alt_mean,
                markersize=4,
                marker="x",
                linestyle="None",
                alpha=0.75,
                color=clr,
            )        
    ax.set_xticks(chroms.Midpoint, labels=chroms.index)
    ax.set_xticks(chroms.Endpoint, minor=True)
    ax.set_xlabel("Coordinates")
    if diff:
        max_abs_diff = max(np.abs(ax.get_ylim()))
        max_x = max(*ax.get_xlim())
        ax.set_ylim(-max_abs_diff, max_abs_diff)
        ax.set_ylabel(r"Alt_mean - Ref_mean")
        ax.hlines(y = 0, xmin = 0, xmax = max_x, linestyle = "dashed", color = "black")
    else:
        ax.set_ylabel(r"Allele-specific means")

    if bool(output):
        plt.savefig(output, format="png", dpi=150)
    else:
        plt.show()




@cli.command()
@click.option(
    "--phenoname",
    "-p",
    type=str,
    default="",
    help=(
        "Name of the phenotype column to analyze. "
        "Defaults to first column other than Sample_Name."
    ),
)
@click.option(
    "--genos",
    type=click.Choice(["01", "012", "-11"]),
    default="01",
    help="Genotype states that are used for analysis.",
)
@click.argument("genotypefile", type=click.Path(exists=True))
@click.argument("phenotypefile", type=click.Path(exists=True))
@click.argument("outfile", type=click.File("w"))
def phenotype(genotypefile, phenotypefile, outfile, phenoname, genos):
    """
    Calculate allele specific means at every variable site.

    GENOTYPES and PHENOTYPES are CSV files. See README for formatting.

    OUTFILE is the name of the CSV file to write that contains coordinates and allele-specific means.
    A dash ('-') can be substituted OUTFILE to write to stdout.
    """
    genodf, phenodf = load_files(genotypefile, phenotypefile)
    if phenoname == "":
        phenoname = phenodf.columns[0]

    if genos == "012":
        genotup = (0, 1, 2)
    elif genos == "-11":
        genotup = (-1, 1)
    else:
        genotup = (0, 1)

    means = means_by_genotype(genodf, phenodf, phenoname, genos=genotup)
    means.to_csv(outfile)        


@cli.command()
@click.option(
    "--phenoname",
    "-p",
    type=str,
    default="",
    help=(
        "Name of the phenotype column to analyze. "
        "Defaults to first column other than Sample_Name."
    ),
)
@click.option(
    "--genos",
    type=click.Choice(["01", "012", "-11"]),
    default="01",
    help="Genotype states that are used for analysis.",
)
@click.option(
    "--test",
    type=click.Choice(assoc_test_dict.keys()),
    default="anova",
    help="Which statistical test of mean differences to apply. Defaults to ANOVA.",
)
@click.argument("genotypefile", type=click.Path(exists=True))
@click.argument("phenotypefile", type=click.Path(exists=True))
@click.argument("outfile", type=click.File("w"))
def stats(genotypefile, phenotypefile, outfile, phenoname, genos, test):
    """
    Calculate association statistics between genotype and phenotype.

    GENOTYPES and PHENOTYPES are CSV files. See README for formatting.

    OUTFILE is the name of the CSV file to write that contains p-values.
    A dash ('-') can be substituted OUTFILE to write to stdout.
    """
    genodf, phenodf = load_files(genotypefile, phenotypefile)
    if phenoname == "":
        phenoname = phenodf.columns[0]

    if genos == "012":
        genotup = (0, 1, 2)
    elif genos == "-11":
        genotup = (-1, 1)
    else:
        genotup = (0, 1)

    assoc = assoc_test_pandarallel(genodf, phenodf, phenoname, genos=genotup, test=test)
    assoc.to_csv(outfile)


def permuted_assoc_max(genodf, phenodf, focalpheno, genos=(0, 1), test="anova", n = 10):
    max_log10 = []
    for i in range(n):
        phenodf[focalpheno] = np.random.permutation(phenodf[focalpheno])
        stats = assoc_test_pandarallel(genodf, phenodf, focalpheno, genos, test)
        max_log10.append(float(stats.log10Pvalue.max()))
    return max_log10

    


@cli.command()
@click.option(
    "--phenoname",
    "-p",
    type=str,
    default="",
    help=(
        "Name of the phenotype column to analyze. "
        "Defaults to first column other than Sample_Name."
    ),
)
@click.option(
    "--genos",
    type=click.Choice(["01", "012", "-11"]),
    default="01",
    help="Genotype states that are used for analysis.",
)
@click.option(
    "--test",
    type=click.Choice(assoc_test_dict.keys()),
    default="anova",
    help="Which statistical test of mean differences to apply. Defaults to ANOVA.",
)
@click.option(
    '--n',
    type=int,
    default=500,
    help="Number of permutations to run for estimating EVD of log10 Pvals"
)
@click.argument("genotypefile", type=click.Path(exists=True))
@click.argument("phenotypefile", type=click.Path(exists=True))
@click.argument("outfile", type=click.File("w"))
def permute(genotypefile, phenotypefile, outfile, phenoname, genos, test, n):
    """
    Calculate maximum log10 Pvals of permuted  datasets.
    """
    genodf, phenodf = load_files(genotypefile, phenotypefile)
    if phenoname == "":
        phenoname = phenodf.columns[0]

    if genos == "012":
        genotup = (0, 1, 2)
    elif genos == "-11":
        genotup = (-1, 1)
    else:
        genotup = (0, 1)

    max_log10pvals = permuted_assoc_max(genodf, phenodf, phenoname, genos=genotup, test=test, n=n)
    maxdf = pd.DataFrame({
        "max_log10Pval":max_log10pvals,
        }
        )
    maxdf.to_csv(outfile, index=False)    



if __name__ == "__main__":
    cli()
